days/day-07/solutions/day07.sml

```diff
@@ -0,0 +1,64 @@
+use "../../../util.sml";
+
+local
+    structure SS = StringSet
+    structure SM = StringMap
+in
+
+fun toDigit c = Char.ord c - Char.ord #"0"
+fun toRule s =
+    let
+        val (parent, rawChildClauses) =
+            case String.substring (s, 0, String.size s - 1)
+                         >> String.fields (fn c => c = #" ")
+                         >> splitWith (fn x => x = "contain") of
+                [a, b] => (String.concatWith " " a, String.concatWith " " b)
+              | _ => raise Fail $ "bad input: " ^ s
+        (* sigh *)
+        fun clause 0 (#" " :: rest) = clause 0 rest
+          | clause 1 (#" " :: name) = SOME (1, (String.implode name) ^ "s")
+          | clause n (#" " :: name) = SOME (n, String.implode name)
+          | clause n (#"n" :: _) = NONE
+          | clause n (x :: rest) = clause (n*10 + toDigit x) rest
+          | clause _ _ = NONE
```

Comment:
Why the `(* Sigh? *)` Is this a switch?

Reply:
It's a function with pattern matching.

The sigh: This is a potpourri of small hacks in the same function. `clause 0 (#" " :: rest)` hacks out the space after a comma (`, `) because Standard ML's standard library makes it hard to split by strings of characters. `clause 1 (#" " :: name)` adds on a `s` to the end to normalise names.. The function itself is there because I can't split by the first space only because Standard ML's standard library.

Reply:
:thinking: